### ISSUE TITLE
SMS data import from Twilio

### ISSUE THREAD

**Issue by HanaShiho** [episphere/connect#886]
For SMS, we'd like to import the event-level message status from the Twilio Webhook. Here's the list of the possible status options, which may be one or multiple variables:

- Accepted 
- Scheduled 
- Queued 
- Sending 
- Sent 
- Delivery Unknown
- Delivered 
- Undelivered, Error code if available 
- Failed, Error code if available 

Note that some event statuses, such as "Sent" or "Sending", will get updated once it's in a final state (e.g., delivered, failed, etc.). 

@jhflorey if you need any help figuring out how these variables are defined and labeled in Twilio, I'm happy to review them with you. Additional resources on tracking the Message Status of outbound msgs can be found [HERE](https://www.twilio.com/docs/messaging/guides/track-outbound-message-status)

Another piece of data under consideration for importing is Time of Delivery (i.e. the time it takes for an SMS to be delivered after being sent). This variable currently does not exist in Twilio UI so we'd need to build logic using the Webhook data or logs to track it. More information to come. 

In terms of data storage, do you want to import them into the SendGrid Tracking collection in Firestore or someplace else?

**Comment by jhflorey** [comment]
@HanaShiho  Do we have the twilio account and register our application in Twilio? The  SendGrid Tracking collection is being used to store the status of sendgrid email , I don't think storing the state of twilio sms  into the  SendGrid Tracking collection is  good idea.

**Comment by HanaShiho** [comment]
@jhflorey Yes, we have a Twilio account and we can access the Twilio UI/dashboard. @we-ai might be able to share more insights as to how it's configured in the backend.

**Comment by HanaShiho** [comment]
@jhflorey Tony was able to get me access. Once you log in to the console, you should be able to see all three accounts (main, stage, dev) and relevant info (e.g. account SIDs, registered phone numbers, etc.). According to Twilio, messaging service features can be managed and configured directly through the console as well as using the REST API.

**Comment by jhflorey** [comment]
I tried integrating our system with twilio message service webhook. As shown below we can see the response after sending (blue block) and the response from the webhook (3 green blocks). It's related each other by `MessageSid: SMd214a5c110e750f63732ee12a7ac77ff`.
![image](https://github.com/episphere/connect/assets/24559107/32358317-1c48-4726-83d4-855a2fac63c2)

I noticed our system stores sms information in `notifications` collection. In `notification` we store notification data including email(sendgrid) and sms(twilio). In the future, we will move all data in `sendgridTracking` to `notifications`. That means we don't need to create `twilioSmsTracking` collection, we should also store twilio webhook data in `notifications` collection. We can define the notification type by the `notificationType` field in the `notifications` collection.

About notification status. We should not override status. I suggest we add more data to the information. Like after sending email we will add `sendingDate` then we have `queuedDate`, `sentDate`, `deliveredDate`. So we will have the time between each status to monitor.

**Comment by jhflorey** [comment]
> In addition to the SMS status variables listed above, we want to track the click-through rate, which is the percentage of recipients who clicked on a link, such as to MyConnect, in the text message.
> 
> We can get the number of delivered messages (denominator) from Webhook, but the number of clicks (numerator) isn't visible in Twilio. So, we are hoping to track clicks via MyConnect or use Google Analytics. @jhflorey, can you tell me which would be feasible to set up?

i believe it's possible. So we will add the query in the link just like `https://myconnect.cancer.gov/?sms=twilio` then we can tracking from the comparison for GA4
<img width="311" alt="image" src="https://github.com/episphere/connect/assets/24559107/1ca31a83-d62a-49c5-b1a4-1b2e98912a90">

**Comment by brotzmanmj** [comment]
@jhflorey If we don't override Status, how do you suggest we store? I agree with storing the status dates.
@FrogGirl1123 can you give your input on this?

If we store it directly in the NotificationTable, that's fine, we just have to be very careful with aligning it with the correct record. The Notifications Table currently only has data on outgoing data, not incoming data. We had issues previously with merging the SendGridTracking data with the NotificationsTable data, i.e. you cannot merge correctly on someone's email or phone number. I believe it needs to be merged on message ID, token, and mode (email or SMS) for it to align correctly.

**Comment by jhflorey** [comment]
Hi @brotzmanmj, The status date determines the status. For example, we have `queuedDate` and `sentDate` and no `diliveredDate`, meaning the message was sent but not yet delivered to the participant.

`The Notifications Table currently only has data on outgoing data, not incoming data` => Looks like it's correct.

` We had issues previously with merging the SendGridTracking data with the NotificationsTable data, i.e. you cannot merge correctly on someone's email or phone number. I believe it needs to be merged on message ID, token, and mode (email or SMS) for it to align correctly` ==> It should be `notification_id` in `sendgridTracking` collection  and `id` in `notifictions` collection.

**Comment by brotzmanmj** [comment]
@we-ai  is the notification_id unique to every message of every type for every person? Is this different than the message ID we looked at yesterday?

@FrogGirl1123 do you have any concerns from an analytics perspective of having just the dates and no flag (meaning analysts will need to derive Status based on the dates and the existence of the dates.

**Comment by we-ai** [comment]
@brotzmanmj Yes, each notification record (email or SMS message) stored in `notifications` collection  has a unique ID. This ID is different from notification spec (schema) ID.

**Comment by brotzmanmj** [comment]
@we-ai Thanks for confirming. So the import strategy sounds good. 

@jhflorey in Dev there are some test SMS messages so you could use those to try to import the data to the Notifications collection. Adding @KELSEYDOWLING7  and @jacobmpeters for their awareness.

For now, please add the dates associated with the SMS message status while we await input from Nicole about the status flag variable. Thanks.

**Comment by jhflorey** [comment]
@brotzmanmj would you like me to create a field that describes the latest status of the sms. Example status = queue or sent or delivered ..... So i think it will help us query ease in big query.

**Comment by brotzmanmj** [comment]
Hi @jhflorey Yes, we were hoping to get a status variable that tell us what is the status of the message that we think has these options: 
Accepted
Scheduled
Queued
Sending
Sent
Delivery Unknown
Delivered
Undelivered, (can we capture the Error code if available?)
Failed (can we capture the Error code if available?)

So will the status update until it is in a final state (delivered, undelivered, failed)? And how often will our system run to capture the latest status)?

**Comment by FrogGirl1123** [comment]
I agree it would be preferable to have a status flag.  Redundancies can be helpful in hunting down issues.  I thought we decided at the in person meeting that we would have a separate table/document for incoming data because of the merge issues @brotzmanmj described.  I am OK with having it in the Notification table as long as there aren't concerns about alignment with the correct message and participant.

**Comment by brotzmanmj** [comment]
Sounds good. 

@jhflorey how often will the flag be updated?
And if you would like to test the merge in dev, there are some SMS messages in there now from testing that you can use and we can confirm.

**Comment by jhflorey** [comment]
@brotzmanmj 
- `how often will the flag be updated?` => It depends on twilio system. Every time twilio returns to our system the new status of the sms, our system will update in firestore.
- `And if you would like to test the merge in dev, there are some SMS messages in there now from testing that you can use and we can confirm.` => We can not test with these old data. Receiving sms status is only effective after we integrate twilio webhook. It means we can only test with new data. I will create new fake data and show it for you guys review it.

**Comment by brotzmanmj** [comment]
@we-ai Please see the first question above. Do you know how often twilio will return data to our system?

@jhflorey  we are continuing to test messages today in dev (will go out at 3pm) and will probably have more testing to do between now and next week. Your fake data will be good, but also would be good to coordinate some testing on this with @HanaShiho to check this for known messages that we are testing.

**Comment by we-ai** [comment]
@brotzmanmj Based on my understanding, data sendings to us from Twilio are triggered by actions: SMS messages delivered, failed, a new message (eg "STOP") coming in, etc. So the frequency of Twilio sending us data depends on how often those actions occur.

**Comment by brotzmanmj** [comment]
OK so it sounds like this is something we will observe as we go

**Comment by jhflorey** [comment]
@brotzmanmj @HanaShiho there are some new fields as the image below after integrating with Twilio WEbhook. Pelase review it. I also remove my fake data in dev

<img width="557" alt="image" src="https://github.com/episphere/connect/assets/24559107/7ff7a421-ec10-4b60-b64c-57607d879d38">

**Comment by brotzmanmj** [comment]
@jhflorey this is all data coming back from Twilio, before it is merged into our Notifications collected?

**Comment by HanaShiho** [comment]
@jhflorey Thanks! Can we also take a look at the Failed and Undelivered msgs to see how that data looks? I'm also curious to see how the error code is being captured. Warren was trying to send a few txt msgs from the stage account but they were undelivered and/or failed so perhaps we can look at those examples?

**Comment by jhflorey** [comment]
will do some investigation on those failed cases later this afternoon.

**Comment by jhflorey** [comment]
> @jhflorey this is all data coming back from Twilio, before it is merged into our Notifications collected?

@brotzmanmj this data is coming back after it is merged into notifications.

**Comment by brotzmanmj** [comment]
Great, @we-ai can you take a look at the data that's merged in? Is it merged on the field called 'id'? I see there is also now 'messageSid' but not sure what that signifies?

**Comment by we-ai** [comment]
In above highlighted fields, `messageSid` is different.  Its value is retrieved from Twilio response when sending a SMS message. It is stored  to `notifications` when the record data are first saved.

**Comment by brotzmanmj** [comment]
Thanks Warren. So can you tell from dev Firestore if the Twilio response data is stored as expected in the Notifications collection for the test Jessica did above?

**Comment by we-ai** [comment]
[Here](https://nih.app.box.com/file/1421729361104) is a file about data structures  of email and SMS message records. In each SMS message record,`messageSid` is retrieved from Twilio response, and it can be used as a unique ID for merging further data updates on the same message from Twilio.

**Comment by brotzmanmj** [comment]
Thank you @we-ai. We can answer any questions from @jhflorey when we meet in the morning.

**Comment by jhflorey** [comment]
@brotzmanmj please help confirm what we need to do for tomorrow dev release.

**Comment by brotzmanmj** [comment]
@jhflorey 

- Update notifications collection with message status data from Twilio
- To correctly align the data, update the notification in the collection that has the same MessageSid
- Include accepted_date, delivered_date, error_code, error_message, queue_date, sent_date, and status
- Get an example of what the data look like for failed messages (Hana has some that failed from earlier testing in dev)

Also: Work with Hana and Rebecca so they can add the new variables from Twilio to the Data Dictionary.

Adding @KELSEYDOWLING7 and @jacobmpeters for their awareness. We hope to have this in dev tomorrow so you can see it in the BQ notifications table in dev next week. Currently the notifications table has data on outgoing messages (emails and SMS). The new data will be the status of each SMS message.

**Comment by jhflorey** [comment]
Hi @HanaShiho  @rebexxxxxx please take a look at this link https://app.box.com/s/ne6xkc5i8gb6qdhxun481bgxjypyz73v for Twilio Data Strcuture & New Fields in `notifications` collection.

**Comment by HanaShiho** [comment]
Thanks @jhflorey! The list of new data fields (on page 7) looks good to me except 'sendingDate.' Sending status is usually only present for a very short time, so this isn't useful for us; please remove it from the list.

**Comment by jhflorey** [comment]
@HanaShiho i've already updated it. Please re-check it.

**Comment by HanaShiho** [comment]
Thanks @jhflorey the updated list looks good.

**Comment by brotzmanmj** [comment]
@jhflorey Hi Jessica, is this in dev now? Can Hana test it out?

**Comment by HanaShiho** [comment]
Upon further discussion, we will only be importing the following SMS status along with the date/time stamps: 

- Delivered
- Undelivered w/ the associated error code and error message 
- Failed with the associated error code and error message

**Comment by jhflorey** [comment]
@brotzmanmj @HanaShiho We will only store fields like the image below

![image](https://github.com/episphere/connect/assets/24559107/199de3a0-7ad3-4ff7-978b-486a7f4a598f)

**Comment by we-ai** [comment]
Can we simplify the 3 dates (`failedDate`, `deliveredDate`, `undeliveredDate`) into 1 (`statusUpdateDate`)? And should we omit `errorCode` and `errorMessage` when there's no error?

**Comment by sonyekere** [comment]
> Can we simplify the 3 dates (`failedDate`, `deliveredDate`, `undeliveredDate`) into 1 (`statusUpdateTime`)? And should we omit `errorCode` and `errorMessage` when there's no error?

Hi! @jacobmpeters , @FrogGirl1123 @brotzmanmj - Warren and Jessica will need confirmation on this data structure update to proceed. Thanks!

**Comment by jhflorey** [comment]
@brotzmanmj @HanaShiho this issue already in dev. We can test it.

**Comment by brotzmanmj** [comment]
@jhflorey Are the data are populating the notifications collection in dev? If so, @HanaShiho can you coordinate with Jessica to test this? Thanks

**Comment by brotzmanmj** [comment]
And @jacobmpeters  and @KELSEYDOWLING7 please check the data in BQ notifications table. You'll need the messageSid of the messages that Hana and Jessica test

**Comment by jacobmpeters** [comment]
@HanaShiho @jhflorey @brotzmanmj Yes, we can can take a look in BQ. Just share the messageSid with us when you are done testing.

**Comment by FrogGirl1123** [comment]
@jacobmpeters and @KELSEYDOWLING7 when you check the data in BQ  could you list here what format the variables are in?  The documentation here, https://nih.app.box.com/file/1446579445728?s=ne6xkc5i8gb6qdhxun481bgxjypyz73v, shows them in as both camelback and snake, it would be good to know which one it is for the DD.  The variables we're adding are in the DD change log now, although were trying to confirm a few fields and asking @brotzmanmj to confirm those are the variables she's expecting.

**Comment by KELSEYDOWLING7** [comment]
Yes, we can compare the data format.  @jacobmpeters Once the testing is complete would you mind rerunning the flattening code on the table?

**Comment by jacobmpeters** [comment]
> Yes, we can compare the data format. @jacobmpeters Once the testing is complete would you mind rerunning the flattening code on the table?

Sure!

**Comment by jhflorey** [comment]
@jacobmpeters @KELSEYDOWLING7 we have 4 messageSid that i tried tomorrow in dev

- SM04adb1dc7aa16289968ab6b6f27222a4 - Failed

- SM1b199ad9333a0e3ee0f4eec58bb95d03 - Failed

- SM8496239d9fde1cc3efe83c486537ad5e - Delivered

- SMec4645c1f58fd12dbbe4a4ad78b8bc35 - Failed

About `Undelivered` i was following Twilio docs as image below,  i tried turning off my phone but still not receiving the event data from Twilio. Keep waiting for the result.

<img width="600" alt="image" src="https://github.com/episphere/connect/assets/24559107/3d2ee822-6821-446b-93cd-0c787cd5ec14">

**Comment by KELSEYDOWLING7** [comment]
failedDate did not populate as expected. It has a FALSE value for all of these scenarios, instead of a time value when the message failed and a null value when the message was delivered

**Comment by HanaShiho** [comment]
failedDate looks correct in Firestore. See the screenshot from Jessica. 
![image](https://github.com/episphere/connect/assets/148134271/e868a8b1-8264-4fba-8174-ab40d3c0a26c)
For the delivered msg, only the deliveredDate is captured so I'm not sure how failedDate should be populated/stored in BQ. 

On a slightly different note, for the Status variable, we might also want to have 'Sent' as a permissible response because I noticed that some messages will remain in the sent status (e.g. unknown delivery), and we might want to capture that information. 

@jhflorey can you please share the messageSid of the msgs we tested this morning?

**Comment by KELSEYDOWLING7** [comment]
@jacobmpeters  @jhflorey  Do we know why it would look different in Firestore then BQ?

**Comment by jacobmpeters** [comment]
@KELSEYDOWLING7 It looks like many of the variables were switched from snake case to camel case since I flattened the data. The ambiguity broke the flattening script. I've resolved the issue and the failedDate now looks appropriate. 

<img width="765" alt="Screenshot 2024-02-22 at 11 41 34 AM" src="https://github.com/episphere/connect/assets/7967451/c39a71e9-4736-4660-840d-bd271a808f0d">

**Comment by KELSEYDOWLING7** [comment]
@jacobmpeters Perfect, thanks!

**Comment by jhflorey** [comment]
@jacobmpeters @KELSEYDOWLING7 @HanaShiho we have new data that i just tested in dev now. 

SM1db177b9c23ba114bde93c8eac52a033 - Delivered
SMca9be0251a597f9788a9521784095fbc - Undelivered
SM408e41b48cd65bdd3f0e729aa6da867b - Failed

It may be available in BQ tomorrow.

**Comment by brotzmanmj** [comment]
@KELSEYDOWLING7  and @jacobmpeters  can you confirm the incoming data from Twilio are aligning accurately with the outgoing data in the Notification data? Meaning, is the incoming data updating the messageSid in a structural way that makes sense to you?

**Comment by KELSEYDOWLING7** [comment]
@brotzmanmj  I believe so. But would you be able to clarify which types of variables are considered outgoing data and which are incoming? I don't see anything that stands out in the variables for these test scenarios, and the variables we have in the Notifications Table prior don't seem to have been altered, but I'm not 100% sure what to be looking for/comparing

**Comment by brotzmanmj** [comment]
HI @KELSEYDOWLING7  So all the data we had in the Notifications table before adding these SMS variables is the outgoing notification data.

**Comment by KELSEYDOWLING7** [comment]
@brotzmanmj  Ok that makes sense. Then yes, all the incoming data aligns with the outgoing data as expected

**Comment by KELSEYDOWLING7** [comment]
Nothing showing up yet in DEV for those most recent 3 test messages. I'll check again on Monday

**Comment by jacobmpeters** [comment]
> @jacobmpeters @KELSEYDOWLING7 @HanaShiho we have new data that i just tested in dev now.
> 
> SM1db177b9c23ba114bde93c8eac52a033 - Delivered SMca9be0251a597f9788a9521784095fbc - Undelivered SM408e41b48cd65bdd3f0e729aa6da867b - Failed
> 
> It may be available in BQ tomorrow.

@jhflorey @brotzmanmj 

@KELSEYDOWLING7 and I independently checked the `Connect.notifications` table in DEV and found no data associated with these tokens. The table was last updated on Feb 25, 2024, 11:03:32 PM UTC-5. So for whatever reason, those data just did not make it to BQ. 

```sql
SELECT
  *
FROM
  nih-nci-dceg-connect-dev.Connect.notifications
WHERE
  token IN ("SM1db177b9c23ba114bde93c8eac52a033",
    "SMca9be0251a597f9788a9521784095fbc",
    "SM408e41b48cd65bdd3f0e729aa6da867b")
```

**Comment by brotzmanmj** [comment]
@jacobmpeters You're saying you don't even have the outgoing Notifications data in the table, and also not the SMS response data from Twilio? Nothing at all for those tokens in BQ? Yet @HanaShiho and @jhflorey we know for sure that the messages did go out? Can we check what it shows in Firestore?

**Comment by jhflorey** [comment]
@jacobmpeters @KELSEYDOWLING7  @brotzmanmj it should be `messageSid`.

**Comment by KELSEYDOWLING7** [comment]
@jhflorey  Oh thank you! My mistake, I'm so used to viewing backend data by token or Connect ID.

Scenarios 1 and 3 look good have the expected incoming and outgoing data. For scenario 2, I didn't see an example here https://nih.app.box.com/file/1446579445728?s=ne6xkc5i8gb6qdhxun481bgxjypyz73v for what to expect for undelivered status. But it does have a status of undelivered, an errorCode of 30005, errorMessage of "Unknown destination handset. The destination number you are trying to reach is unknown and may no longer exist." and failedDate is null.

**Comment by brotzmanmj** [comment]
@KELSEYDOWLING7 Just so I understand, you found them based on messageSid but the messageSid is associated with a person right (it should exist in relation to a token). Is that correct?

**Comment by jhflorey** [comment]
@KELSEYDOWLING7 we have `undeliveredDate` instead of `failedDate` for undelivered case.

**Comment by KELSEYDOWLING7** [comment]
@jhflorey  Perfect thank you, they do have an undelivered date.

@brotzmanmj  That's correct. I found them based on the messageSID Jessica listed and they do each have a token

**Comment by brotzmanmj** [comment]
Great! Sounds like this is ready for stage

**Comment by jhflorey** [comment]
> I was given these two messageSID from Hana with the status of delivered and they match in BQ SM5f2a77a0bcc0aa22a315040985644fb9 SMbe239a0c4540d47e9026bb0c3674c182

@KELSEYDOWLING7 it's from stage, right?

**Comment by KELSEYDOWLING7** [comment]
@jhflorey  Yup!
